File: goldenflow/transforms/auto_correct.py
from __future__ import annotations

from collections import Counter

import polars as pl
from rapidfuzz import fuzz

from goldenflow.transforms import register_transform


def _build_canonical_map(
    values: list[str | None],
    frequency_threshold: float = 0.05,
    match_threshold: float = 85.0,
) -> dict[str, str]:
    """Build a mapping from variant values to canonical values.

    1. Count case-insensitive frequencies
    2. Values above frequency_threshold are canonical candidates
    3. Low-frequency values fuzzy-matched against candidates
    """
    # Count case-insensitive frequencies
    lower_counts: Counter[str] = Counter()
    case_map: dict[str, Counter[str]] = {}  # lowercase -> Counter of original casings

    for v in values:
        if v is None:
            continue
        v_stripped = v.strip()
        if not v_stripped:
            continue
        low = v_stripped.lower()
        lower_counts[low] += 1
        if low not in case_map:
            case_map[low] = Counter()
        case_map[low][v_stripped] += 1

    total = sum(lower_counts.values())
    if total == 0:
        return {}

    # Determine canonical values: high-frequency case-insensitive groups
    # A value is canonical if its relative frequency exceeds the threshold
    canonical_values: dict[str, str] = {}  # lowercase -> most common casing
    low_freq_values: list[str] = []

    for low, count in lower_counts.items():
        if (count / total) >= frequency_threshold:
            # Most common casing is the canonical form
            canonical_values[low] = case_map[low].most_common(1)[0][0]
        else:
            low_freq_values.append(low)

    # Fuzzy match low-frequency values against canonical ones
    corrections: dict[str, str] = {}

    # First: exact case-insensitive matches (e.g., "ACTIVE" -> "active")
    for low in list(canonical_values.keys()):
        for original_casing in case_map[low]:
            if original_casing != canonical_values[low]:
                corrections[original_casing] = canonical_values[low]

    # Second: fuzzy matches for truly different strings
    canonical_list = list(canonical_values.keys())
    for low in low_freq_values:
        best_score = 0.0
        best_match = ""
        for canon_low in canonical_list:
            score = fuzz.ratio(low, canon_low)
            if score > best_score:
                best_score = score
                best_match = canon_low
        if best_score >= match_threshold:
            # Map all original casings of this low-freq value to the canonical
            for original_casing in case_map[low]:
                corrections[original_casing] = canonical_values[best_match]

    return corrections


@register_transform(
    name="category_auto_correct",
    input_types=["string"],
    auto_apply=True,
    priority=35,
    mode="series",
)
def category_auto_correct(
    series: pl.Series,
    frequency_threshold: float = 0.05,
    match_threshold: float = 85.0,
) -> pl.Series:
    """Auto-correct categorical misspellings and case variants.

    Identifies low-frequency values that fuzzy-match high-frequency canonical
    values and corrects them. No LLM required.
    """
    values = series.to_list()
    corrections = _build_canonical_map(values, frequency_threshold, match_threshold)

    if not corrections:
        return series

    def _correct(val: str | None) -> str | None:
        if val is None:
            return None
        v_stripped = val.strip()
        return corrections.get(v_stripped, v_stripped)

    return series.map_elements(_correct, return_dtype=pl.Utf8)

File: packages/goldenflow-js/src/core/transforms/auto-correct.ts
/**
 * Auto-correct transform — ported from goldenflow/transforms/auto_correct.py
 * Side-effect module: registers the category_auto_correct transform on import.
 *
 * Uses Levenshtein-based fuzzy matching to correct low-frequency categorical
 * values to their most likely canonical form.
 */

import type { ColumnValue } from "../types.js";
import { registerTransform } from "./registry.js";

// ---------------------------------------------------------------------------
// Levenshtein distance + fuzzy ratio
// ---------------------------------------------------------------------------

function levenshtein(a: string, b: string): number {
  const m = a.length;
  const n = b.length;

  if (m === 0) return n;
  if (n === 0) return m;

  // Use single-row DP for space efficiency
  const prev = new Array<number>(n + 1);
  for (let j = 0; j <= n; j++) prev[j] = j;

  for (let i = 1; i <= m; i++) {
    let prevDiag = prev[0]!;
    prev[0] = i;
    for (let j = 1; j <= n; j++) {
      const temp = prev[j]!;
      if (a[i - 1] === b[j - 1]) {
        prev[j] = prevDiag;
      } else {
        prev[j] = 1 + Math.min(prevDiag, prev[j]!, prev[j - 1]!);
      }
      prevDiag = temp;
    }
  }

  return prev[n]!;
}

/**
 * Fuzzy similarity ratio between two strings (0-100).
 * 100 means identical, 0 means completely different.
 */
function fuzzyRatio(a: string, b: string): number {
  if (a.length === 0 && b.length === 0) return 100;
  const maxLen = Math.max(a.length, b.length);
  const dist = levenshtein(a, b);
  return 100 * (1 - dist / maxLen);
}

// ---------------------------------------------------------------------------
// category_auto_correct (series, string, 35, auto_apply)
// ---------------------------------------------------------------------------

function categoryAutoCorrect(
  values: readonly ColumnValue[],
  frequencyThreshold: unknown = 0.05,
  matchThreshold: unknown = 85,
): ColumnValue[] {
  const freqThresh =
    typeof frequencyThreshold === "number"
      ? frequencyThreshold
      : Number(frequencyThreshold) || 0.05;
  const matchThresh =
    typeof matchThreshold === "number"
      ? matchThreshold
      : Number(matchThreshold) || 85;

  // 1. Count case-insensitive frequencies
  const freqMap = new Map<string, number>(); // lowercase -> count
  const casingMap = new Map<string, Map<string, number>>(); // lowercase -> (original -> count)
  let totalNonNull = 0;

  for (const v of values) {
    if (v === null || typeof v !== "string") continue;
    const lower = v.toLowerCase();
    totalNonNull++;
    freqMap.set(lower, (freqMap.get(lower) ?? 0) + 1);

    let casings = casingMap.get(lower);
    if (!casings) {
      casings = new Map<string, number>();
      casingMap.set(lower, casings);
    }
    casings.set(v, (casings.get(v) ?? 0) + 1);
  }

  if (totalNonNull === 0) return values.slice();

  // 2. Determine canonical candidates: values above frequency threshold
  //    Use the most common casing for each canonical
  const canonicals = new Map<string, string>(); // lowercase -> best casing

  for (const [lower, count] of freqMap) {
    if (count / totalNonNull >= freqThresh) {
      // Pick the most common casing
      const casings = casingMap.get(lower)!;
      let bestCasing = lower;
      let bestCount = 0;
      for (const [original, c] of casings) {
        if (c > bestCount) {
          bestCount = c;
          bestCasing = original;
        }
      }
      canonicals.set(lower, bestCasing);
    }
  }

  if (canonicals.size === 0) return values.slice();

  // 3. Build correction map for low-frequency values
  const corrections = new Map<string, string>(); // lowercase -> canonical

  for (const [lower] of freqMap) {
    if (canonicals.has(lower)) continue; // Already canonical

    let bestCanonical: string | null = null;
    let bestScore = 0;

    for (const [canonLower, canonOriginal] of canonicals) {
      const score = fuzzyRatio(lower, canonLower);
      if (score >= matchThresh && score > bestScore) {
        bestScore = score;
        bestCanonical = canonOriginal;
      }
    }

    if (bestCanonical !== null) {
      corrections.set(lower, bestCanonical);
    }
  }

  // 4. Apply corrections
  return values.map((v) => {
    if (v === null || typeof v !== "string") return v;
    const lower = v.toLowerCase();
    const correction = corrections.get(lower);
    if (correction !== undefined) return correction;
    // If it's a canonical, normalize to best casing
    const canonical = canonicals.get(lower);
    if (canonical !== undefined) return canonical;
    return v;
  });
}

registerTransform(
  { name: "category_auto_correct", inputTypes: ["string"], autoApply: true, priority: 35, mode: "series" },
  categoryAutoCorrect,
);
